File: 2022/PythonSolutions/Day12.py
from collections import namedtuple, deque
from multiprocessing import Pool


Pos = namedtuple('Pos', 'r c count')
def find_path(board, end, queue, visited):
    while queue:
        pos = queue.popleft()
        for dr in ((0, 1), (0, -1), (1, 0), (-1, 0)):
            try:
                new_pos = (pos.r + dr[0], pos.c + dr[1])
                if new_pos not in visited and board[new_pos[0]][new_pos[1]] - board[pos.r][pos.c] < 2:
                    if new_pos == end:
                        return pos.count + 1
                    queue.append(Pos(new_pos[0], new_pos[1], pos.count + 1))
                    visited.add(new_pos)
            except IndexError: pass
    return float("inf")

with open("../Inputs/InputDay12.txt") as file:
    board = [[ord(x) for x in line.rstrip()] for line in file]
    start = end = None
    starts = []
    for r, row in enumerate(board):
        for c, val in enumerate(row):
            if val == ord('S'):
                board[r][c], start = ord('a'), Pos(r, c, 0)
            elif val == ord('E'):
                board[r][c], end = ord('z'), (r, c)
            elif val == ord('a'):
                starts.append(Pos(r, c, 0))
    print(find_path(board, end, deque([start]), {(start.r, start.c)}))

    with Pool() as pool:
        results = [pool.apply_async(find_path, (board, end, deque([s]), {(s.r, s.c)})) for s in starts]
        print(min(p.get() for p in results))
    print(int("dsg"))

File: 2022/rust_solutions/src/day12.rs
use std::collections::{HashSet, VecDeque};
use std::sync::{mpsc, Arc};
use std::{fs, thread};

const PART1: bool = false;

pub fn solve() {
    let data = fs::read_to_string("../Inputs/InputDay12.txt").expect("Unable to read file");
    let mut board: Vec<Vec<usize>> = data
        .split('\n')
        .map(|line| line.chars().map(|x| x as usize).collect())
        .collect();
    let mut start = Pos::default();
    let mut end = (0, 0);
    let mut starts = vec![];
    for (r, row) in board.iter().enumerate() {
        for (c, val) in row.iter().enumerate() {
            if *val == 'S' as usize {
                start = Pos { r, c, count: 0 };
            } else if *val == 'E' as usize {
                end = (r, c);
            } else if *val == 'a' as usize {
                starts.push(Pos { r, c, count: 0 })
            }
        }
    }
    board[start.r][start.c] = 'a' as usize;
    board[end.0][end.1] = 'z' as usize;

    let board_arc = Arc::new(board);

    let res = find_path(
        board_arc.clone(),
        end,
        VecDeque::from([start.clone()]),
        HashSet::from([(start.r, start.c)]),
    );
    println!("part1: {res}");

    let mut rxs = vec![];
    for start in starts {
        let (tx, rx) = mpsc::channel();
        let arc = board_arc.clone();
        thread::spawn(move || {
            let res = find_path(
                arc,
                end,
                VecDeque::from([start.clone()]),
                HashSet::from([(start.r, start.c)]),
            );
            tx.send(res).unwrap();
        });
        rxs.push(rx);
    }
    let res = rxs.iter().filter_map(|x| x.recv().ok()).min().unwrap();
    println!("part2: {res}")
}

#[derive(Default, Clone)]
struct Pos {
    r: usize,
    c: usize,
    count: usize,
}

fn find_path(
    board: Arc<Vec<Vec<usize>>>,
    end: (usize, usize),
    mut queue: VecDeque<Pos>,
    mut visited: HashSet<(usize, usize)>,
) -> usize {
    while let Some(pos) = queue.pop_front() {
        for dr in [(0, 1), (0, -1), (1, 0), (-1, 0)] {
            let new_pos = (pos.r as isize + dr.0, pos.c as isize + dr.1);
            if new_pos.0 < 0
                || new_pos.0 >= board.len() as isize
                || new_pos.1 < 0
                || new_pos.1 >= board[0].len() as isize
            {
                continue;
            }
            let new_pos = (new_pos.0 as usize, new_pos.1 as usize);
            if !visited.contains(&new_pos) && board[new_pos.0][new_pos.1] < board[pos.r][pos.c] + 2
            {
                if new_pos == end {
                    return pos.count + 1;
                }
                queue.push_back(Pos {
                    r: new_pos.0,
                    c: new_pos.1,
                    count: pos.count + 1,
                });
                visited.insert(new_pos);
            }
        }
    }
    usize::MAX
}
